geti_sdk/post_inference_hooks/post_inference_hook.py

```diff
@@ -0,0 +1,171 @@
+        trigger is activated.
+
+    By default, hooks are executed in parallel in a separate thread from the main
+    inference code. If you want to disable parallel execution, make sure to set
+    `max_threads=0` when creating the hook.
+
+    :param trigger: PostInferenceTrigger which is evaluated and used to trigger the
+        hook action
+    :param action: PostInferenceAction to be executed if triggered
+    :param max_threads: Maximum number of threads to use for hook execution. Defaults
+        to 5. Set to 0 to disable parallel hook execution.
+    :param limit_action_rate: True to limit the rate at which the `action` for the
+        hook is executed. Note that rate limiting is applied **after** trigger
+        evaluation, so this can be useful for handling video streaming in which
+        subsequent frames may be very similar. Enabling rate limiting will avoid
+        sampling many near-duplicate frames
+    :param max_frames_per_second: Maximum frame rate at which to execute the action.
+        Only takes effect if `limit_action_rate = True`.
+    """
+
+    def __init__(
+        self,
+        trigger: PostInferenceTrigger,
+        action: PostInferenceAction,
+        max_threads: int = 5,
+        limit_action_rate: bool = False,
+        max_frames_per_second: float = 1,
+    ):
+        super().__init__(trigger=trigger, action=action)
+
+        self.parallel_execution = max_threads != 0
+        if self.parallel_execution:
+            self.executor = ThreadPoolExecutor(max_workers=max_threads)
+            logging.debug(
+                f"Parallel inference hook execution enabled, using a maximum of "
+                f"{max_threads} threads."
+            )
+
+        if limit_action_rate:
+            self.rate_limiter: Optional[RateLimiter] = RateLimiter(
+                frames_per_second=max_frames_per_second, is_blocking=False
+            )
+        else:
+            self.rate_limiter = None
+
+    def run(
+        self,
+        image: np.ndarray,
+        prediction: Prediction,
+        name: Optional[str] = None,
+        timestamp: Optional[datetime.datetime] = None,
+    ) -> None:
+        """
+        Run the post inference hook. First evaluate the trigger, then execute the
+        action if the trigger conditions are met.
+
+        :param image: Numpy array representing the image
+        :param prediction: Prediction object containing the inference results for the
+            image
+        :param name: Optional name of the image which can be used in the hook
+            action, for example as filename or tag for data collection
+        :param timestamp: Optional timestamp belonging to the image
+        """
+
+        def execution_function(
+            im: np.ndarray, pred: Prediction, ts: datetime.datetime
+        ) -> None:
+            score = self.trigger(image=im, prediction=pred)
+            decision = self.trigger.get_decision(score=score)
```

Comment:
I don't think so because we need both the decision and the score: The decision determines whether or not to execute the action, and the score needs to be passed to the action. So I think it's fine to have it in 2 lines like this